pkg/kine/prepared/db.go

```diff
@@ -0,0 +1,191 @@
+package prepared
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+type DB struct {
+	underlying *sql.DB
+	mu         sync.Mutex
+	maxSize    int
+	store      map[string]*lruEntry
+	lruList    lruList
+}
+
+func New(db *sql.DB, maxSize int) *DB {
+	return &DB{
+		underlying: db,
+		maxSize:    maxSize,
+		store:      make(map[string]*lruEntry, maxSize),
+	}
+}
+
+func (db *DB) Underlying() *sql.DB { return db.underlying }
+
+func (db *DB) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.ExecContext(ctx, args...)
+}
+
+func (db *DB) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.QueryContext(ctx, args...)
+}
+
+func (db *DB) Close() error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	errs := []error{}
+	for len(db.store) != 0 {
+		if err := db.drop(db.lruList.tail()); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := db.underlying.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (db *DB) touch(entry *lruEntry) {
+	db.lruList.remove(entry)
+	db.lruList.add(entry)
+}
+
+func (db *DB) prepare(ctx context.Context, query string) (*stmt, error) {
+	if stmt := db.get(query); stmt != nil {
+		return stmt, nil
+	}
+	prepared, err := db.underlying.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	result := &stmt{prepared, atomic.Int32{}}
+	result.Ref()
+
+	db.put(query, result)
+	return result, nil
+}
+
+func (db *DB) get(key string) *stmt {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	if entry, ok := db.store[key]; ok {
+		db.touch(entry)
+		return entry.stmt.Ref()
+	}
+	return nil
+}
+
+func (db *DB) put(query string, stmt *stmt) {
+	if db.maxSize <= 0 {
+		return
+	}
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	entry, ok := db.store[query]
+	if ok {
+		db.touch(entry)
+		entry.stmt.Close()
+		entry.stmt = stmt.Ref()
+		return
+	}
+
+	if len(db.store) >= db.maxSize {
+		db.drop(db.lruList.tail())
```

Comment:
I think if the cache were moved into its own struct, we could add our own method on it to make it clear what we're getting here (somehow the 'tail' of the LRU list being the least recently used and not the most recently used seems like an implementation detail). I think it would be really nice if this line read something like
```go
db.drop(db.cache.LeastRecentlyUsed())
```
Also, I think there's either some missing error propagation here, or, if we're ignoring errors at this point (which seems fairly reasonable), it would be good to explicitly state that we're ignoring them, e.g. with
 ```go
_ = db.drop(...)
```
I actually wonder whether this is the right place to ignore these errors, or whether they should be passed upwards to be ignored elsewhere?

Reply:
Removed.